tests/integration/ha/test_ha.py

```diff
@@ -0,0 +1,491 @@
+    )
+    await asyncio.sleep(CLIENT_TIMEOUT * 3)  # letting client set up and start writing
+
+    logger.info("Checking writes are running at all...")
+    assert cw.count_znodes(
+        parent=parent, hosts=hosts, username=helpers.USERNAME, password=password
+    )
+
+    logger.info("Stopping leader process...")
+    await helpers.send_control_signal(ops_test=ops_test, unit_name=leader_name, signal="SIGSTOP")
+    await asyncio.sleep(CLIENT_TIMEOUT * 3)  # to give time for re-election
+
+    assert await helpers.process_stopped(ops_test, leader_name)
+
+    logger.info("Checking writes are increasing...")
+    writes = cw.count_znodes(
+        parent=parent, hosts=non_leader_hosts, username=helpers.USERNAME, password=password
+    )
+    await asyncio.sleep(CLIENT_TIMEOUT * 3)  # increasing writes
+    new_writes = cw.count_znodes(
+        parent=parent, hosts=non_leader_hosts, username=helpers.USERNAME, password=password
+    )
+    assert new_writes > writes, "writes not continuing to ZK"
+
+    logger.info("Checking leader re-election...")
+    new_leader_name = helpers.get_leader_name(ops_test, non_leader_hosts)
+    assert new_leader_name != leader_name
+
+    logger.info("Continuing leader process...")
+    await helpers.send_control_signal(ops_test=ops_test, unit_name=leader_name, signal="SIGCONT")
+    await asyncio.sleep(CLIENT_TIMEOUT * 3)  # letting writes continue while unit rejoins
+
+    logger.info("Stopping continuous_writes...")
+    cw.stop_continuous_writes()
+    await asyncio.sleep(CLIENT_TIMEOUT)  # buffer to ensure writes sync
+
+    logger.info("Counting writes on surviving units...")
+    last_write = cw.get_last_znode(
+        parent=parent, hosts=non_leader_hosts, username=helpers.USERNAME, password=password
+    )
+    total_writes = cw.count_znodes(
+        parent=parent, hosts=non_leader_hosts, username=helpers.USERNAME, password=password
+    )
+    assert last_write == total_writes
+
+    logger.info("Checking old leader caught up...")
+    last_write_leader = cw.get_last_znode(
+        parent=parent, hosts=leader_host, username=helpers.USERNAME, password=password
+    )
+    total_writes_leader = cw.count_znodes(
+        parent=parent, hosts=leader_host, username=helpers.USERNAME, password=password
+    )
+    assert last_write == last_write_leader
+    assert total_writes == total_writes_leader
+
+
+@pytest.mark.abort_on_fail
+async def test_full_cluster_crash(ops_test: OpsTest, request):
+    hosts = helpers.get_hosts(ops_test)
+    leader_name = helpers.get_leader_name(ops_test, hosts)
+    leader_host = helpers.get_unit_host(ops_test, leader_name)
+    password = helpers.get_super_password(ops_test)
+    parent = request.node.name
+    non_leader_hosts = ",".join([host for host in hosts.split(",") if host != leader_host])
+
+    logger.info("Extending pebble restart delay on all units...")
+    helpers.modify_pebble_restart_delay(ops_test, policy="extend")
```

Comment:
**suggestion**: I would keep this as a fixture, same as the [`restart_delay` here](https://github.com/canonical/zookeeper-operator/blob/main/tests/integration/ha/test_ha.py#L26)

Reply:
Agreed, in another PR though.